### PR TITLE
BUG: TDI.insert with empty TDI raising IndexError

### PR DESCRIPTION
**Pull request by jbrockmendel**
This started out as a cosmetic-only branch and ended up finding a broken corner case.  The relevant change is in timedeltas L416 where `if self.freq is not None:` is now `if self.size and self.freq is not None:`

Using _check_compatible_with causes us to raise TypeError instead of ValueError in a couple of the DatetimeIndex.insert cases.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
@simonjayhawkins suggestions for fixing mypy?

**Comment by jbrockmendel** [comment]
```
mypy 0.730
Performing static analysis using mypy
pandas/core/indexes/timedeltas.py:56: error: Definition of "_data" in base class "ExtensionIndex" is incompatible with definition in base class "DatetimelikeDelegateMixin"
pandas/core/indexes/timedeltas.py:56: error: Definition of "_data" in base class "Index" is incompatible with definition in base class "DatetimelikeDelegateMixin"
pandas/core/indexes/datetimes.py:87: error: Definition of "_data" in base class "ExtensionIndex" is incompatible with definition in base class "DatetimelikeDelegateMixin"
pandas/core/indexes/datetimes.py:87: error: Definition of "_data" in base class "Index" is incompatible with definition in base class "DatetimelikeDelegateMixin"
Found 4 errors in 2 files (checked 844 source files)
```

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins suggestions for fixing mypy?

am AFK for the rest of the day (Pydata London). will look tomorrow.

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ class DatetimeIndexOpsMixin(ExtensionIndex, ExtensionOpsMixin):
     Common ops mixin to support a unified interface datetimelike Index.
     """
 
-    _data: ExtensionArray
+    _data: Union[DatetimeArray, TimedeltaArray, PeriodArray]
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: should define this with a short label, maybe DatetimeArrayType

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -946,7 +945,7 @@ def insert(self, loc, item):
                     freq = self.freq
                 elif (loc == len(self)) and item - self.freq == self[-1]:
                     freq = self.freq
-            item = _to_M8(item, tz=self.tz)
+            item = item.asm8
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: can we remove _to_M8?

File: pandas/tests/indexes/timedeltas/test_indexing.py [hunk 3]
```diff
@@ -161,6 +161,15 @@ def test_take_fill_value(self):
...
 class TestTimedeltaIndex:
+    def test_insert_empty(self):
+        # Corner case inserting with length zero doesnt raise IndexError
+        idx = timedelta_range("1 Day", periods=3)
+        td = idx[0]
+
+        idx[:0].insert(0, td)
+        idx[:0].insert(1, td)
+        idx[:0].insert(-1, td)
+
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_indexing.py]: maybe worth comparing the results

**Review by jreback — APPROVED** [review]
some possible followup comments, otherwise lgtm.

**Comment by jreback** [comment]
ping on green.

File: pandas/core/indexes/datetimes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: > can we remove _to_M8?

IIRC there's still one more usage to get rid of after this.  I'm eager to get rid of it, as im pretty sure it is hiding a bunch of corner case bugs.

**Comment by jbrockmendel** [comment]
ping

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**